externalclock/clock.go

```diff
@@ -15,29 +15,49 @@ import (
 )
 
 type Clock struct {
-	Logger           logr.Logger
-	timestampChan    chan time.Time // synchronization SetTimestamp -> Run
-	timestampUpdated chan struct{}  // synchronization Run -> SetTimestamp
-	timeMutex        sync.Mutex
-	currentTime      time.Time
-	tickerMutex      sync.RWMutex
-	tickers          map[string]*ticker
+	Logger      logr.Logger
+	timeMutex   sync.Mutex
+	currentTime time.Time
+	tickerMutex sync.RWMutex
+	tickers     map[string]*ticker
 }
 
 func New(logger logr.Logger, initialTime time.Time) *Clock {
 	c := &Clock{
-		Logger:           logger,
-		timestampChan:    make(chan time.Time),
-		tickers:          map[string]*ticker{},
-		timestampUpdated: make(chan struct{}),
+		Logger:  logger,
+		tickers: map[string]*ticker{},
 	}
 	c.currentTime = initialTime
 	return c
 }
 
 func (g *Clock) SetTimestamp(t time.Time) {
-	g.timestampChan <- t
-	<-g.timestampUpdated // wait for time to be set before returning
+	g.timeMutex.Lock()
+	g.currentTime = t
+	g.timeMutex.Unlock()
+
+	g.signalTickers(t)
+}
+
+func (g *Clock) signalTickers(t time.Time) {
+	g.tickerMutex.RLock()
+	for _, tickerInstance := range g.tickers {
+		if !tickerInstance.IsDurationReached(t) {
+			continue
+		}
+		tickerInstance.SetLastTimestamp(t)
+		if !tickerInstance.isPeriodic {
+			g.tickerMutex.RUnlock()
+			tickerInstance.Stop()
+			g.tickerMutex.RLock()
+		}
+
+		tickerInstance := tickerInstance
+		go func() {
+			tickerInstance.timeChan <- t
+		}()
+	}
+	g.tickerMutex.RUnlock()
```

Comment:
Is there a consequence of unlocking the ticker mutex early due to spinning up new routines here?

Reply:
Perhaps, I added a buffer on the time channel, so we don't need a goroutine to send it, so crisis adverted!